### PR TITLE
[Lang] [IR] Add coordinate offset argument for ti.var/Vector/Matrix (#1081)

### PR DESCRIPTION
**Pull request by Rullec**
Hi All,

This PR is **still about coordinate offset args in ti.var and ti.Matrix.** 

The previous PR was unexpectedly closed in the afternoon. So I manually copy all @yuanming-hu 's advice and integrated them into this new PR. Also, under the kind help of @archibate , I finally managed to fix my messed up git. Thanks a lot!

In this PR, I failed to support offset for ti.Vector as @yuanming-hu  advised before. Generally, It's not as simple as I imagined. I first run
```
a = ti.Vector(dt=ti.i32, shape=32) # works well
```
and it works well. But when I tried to add offset in this line of code, an exception was thrown out
```
b = ti.Vector(dt=ti.i32, shape=32, offset = -16) 
```
the traceback log is shown below:
```
[Taichi] mode=development
[Taichi] preparing sandbox at /var/folders/ry/wq5tj2w52793sf91z8qr2dz00000gn/T/taichi-ka7p1ykt
[Taichi] sandbox prepared
[Taichi] <dev mode>, supported archs: [cpu, metal], commit 29b2c840, python 3.6.7
Traceback (most recent call last):
File "test_offset.py", line 81, in <module>
    b = ti.Vector(dt=ti.i32, shape=32, offset = -16) # exception thrown out
File "/Users/xudong/Projects/taichi/python/taichi/lang/matrix.py", line 616, in Vector
    return Matrix(n, 1, dt, shape, offset, **kwargs)
File "/Users/xudong/Projects/taichi/python/taichi/lang/matrix.py", line 128, in __init__
    ti.root.dense(ti.index_nd(dim), shape).place(*tuple(var_list), offset=offset)
File "/Users/xudong/Projects/taichi/python/taichi/lang/snode.py", line 45, in place
    raise ValueError(f'{arg} cannot be placed')
ValueError: None cannot be placed
```
So this feature is still incomplete, and this problem seems to be more low-level (at lease, the SNode class was involved) than my toy modification. Maybe I need some more time for analyzing and digging in it.

Can I submit this PR now? I have no idea about whether I need to fix them all - then submit a single PR. Please give me some advice :D Thank you in advance!

Bests,
Xudong

### CODE REVIEW

File: python/taichi/lang/matrix.py [hunk 1]
```diff
@@ -19,6 +19,7 @@ def __init__(self,
                  dt=None,
                  shape=None,
                  empty=False,
+                 offset=None,
```
Comment (archibate) [inline python/taichi/lang/matrix.py:22]:
```suggestion
                 offset=None,
                 empty=False,
```

File: python/taichi/lang/matrix.py [hunk 2]
```diff
@@ -602,8 +618,8 @@ def outer_product(self, b):
         return c
 
 
-def Vector(n=1, dt=None, shape=None, **kwargs):
-    return Matrix(n, 1, dt, shape, **kwargs)
+def Vector(n=1, dt=None, shape=None, offset=None, **kwargs):
+    return Matrix(n, 1, dt, shape, offset, **kwargs)
```
Comment (yuanming-hu) [inline python/taichi/lang/matrix.py:622]:
```suggestion
    return Matrix(n, 1, dt=dt, shape=shape, offset=offset, **kwargs)
```

File: tests/python/test_offset.py [hunk 3]
```diff
@@ -49,3 +49,29 @@ def test():
...
+
+@ti.all_archs
+def test_offset_for_var():
+    a = ti.var(dt=ti.i32, shape=16, offset=-48)
+    b = ti.var(dt=ti.i32, shape=(16, ), offset=(16, ))
+    c = ti.var(dt=ti.i32, shape=(16, 64), offset=(-16, -64))
+    d = ti.var(dt=ti.i32, shape=(16, 64), offset=None)
+
+
+@ti.must_throw(AssertionError)
```
Comment (yuanming-hu) [inline tests/python/test_offset.py:68]: Should it be `ValueError`?

File: tests/python/test_offset.py [hunk 4]
```diff
@@ -49,3 +49,29 @@ def test():
...
+def test_offset_must_throw_var():
+    a = ti.var(dt=ti.float32, shape=3, offset=(3, 4))
+    b = ti.var(dt=ti.float32, shape=None, offset=(3, 4))
+
+
+@ti.must_throw(AssertionError)
```
Comment (yuanming-hu) [inline tests/python/test_offset.py:74]: `ValueError`?

File: docs/offset.rst [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+Coordinate offsets
+========
+
+- A Taichi tensor can be defined with **coordinate offsets**. The offsets will move tensor bounds so that tensor origins are no longer zero vectors. A typical use case is to support voxels with negative coordinates in physical simulations.
+- For example, a matrix of ``32x64`` elements with coordinate offset ``(-16, 8)`` can be defined as the following:
+
+.. code-block:: python
+
+    a = ti.Matrix(2, 2, dt=ti.f32, shape=(32, 64), offset = (-16, 8))
```
Comment (yuanming-hu) [inline docs/offset.rst:11]:
```suggestion
    a = ti.Matrix(2, 2, dt=ti.f32, shape=(32, 64), offset=(-16, 8))
```

File: tests/python/test_offset.py (hunk 4 above)
Comment (Rullec) [reply tests/python/test_offset.py:74]: Yes, Now I change all these ValueError to AssertError

File: tests/python/test_offset.py [hunk 6]
```diff
@@ -49,3 +49,29 @@ def test():
...
+    c = ti.var(dt=ti.i32, shape=(16, 64), offset=(-16, -64))
+    d = ti.var(dt=ti.i32, shape=(16, 64), offset=None)
+
+
+@ti.must_throw(AssertionError)
+def test_offset_must_throw_var():
+    a = ti.var(dt=ti.float32, shape=3, offset=(3, 4))
+    b = ti.var(dt=ti.float32, shape=None, offset=(3, 4))
+
+
```
[the remark this answers is not in the document]
Comment (Rullec) [reply tests/python/test_offset.py:73]: Yes, the test for Vector has been added now.

File: tests/python/test_offset.py (hunk 3 above)
Comment (Rullec) [reply tests/python/test_offset.py:68]: Yes, it should be indeed. Just now I convert all ValueError to AssertError, would it be ok?

**Comment by Rullec** [comment]
Thank you for these greate suggestions:)!
Just now I do pull this PR pesudo branch to my local end by running
```
git fetch upstream pull/1102/head:pr1102
git checkout pr37
```
and add a offset test for ti.Vector.

Something new happened. When the ti.Vector and ti.Matrix with offset is initalized, SNode reports an error:
```

@ti.all_archs
def test_offset_for_vector():
    offset = 16
    shape = 16
    a = ti.Vector(dt=ti.i32, shape=shape, offset=offset)
    

    @ti.kernel
    def test():
        for i in a:
            a[i] = 2 * i

    test()
    for i in range(offset, offset+shape, 1):
        assert e[i] == 2 * i
    return
```
traceback shown below
```
[Taichi] mode=development
[Taichi] preparing sandbox at /var/folders/ry/wq5tj2w52793sf91z8qr2dz00000gn/T/taichi-dcl63yu2
[Taichi] sandbox prepared
[Taichi] <dev mode>, supported archs: [cpu, metal], commit 29b2c840, python 3.6.7
Running test on arch=Arch.x64
Traceback (most recent call last):
  File "test_offset.py", line 131, in <module>
    test_offset_for_vector()
  File "/Users/xudong/Projects/taichi/python/taichi/lang/__init__.py", line 365, in wrapped
    test(*test_args, **test_kwargs)
  File "test_offset.py", line 89, in test_offset_for_vector
    test()
  File "/Users/xudong/Projects/taichi/python/taichi/lang/kernel.py", line 533, in wrapped
    return primal(*args, **kwargs)
  File "/Users/xudong/Projects/taichi/python/taichi/lang/kernel.py", line 463, in __call__
    self.materialize(key=key, args=args, arg_features=arg_features)
  File "/Users/xudong/Projects/taichi/python/taichi/lang/kernel.py", line 343, in materialize
    taichi_kernel = taichi_kernel.define(taichi_ast_generator)
  File "/Users/xudong/Projects/taichi/python/taichi/lang/kernel.py", line 340, in taichi_ast_generator
    compiled()
  File "test_offset.py", line 87, in test
    a[i] = 2 * i
  File "/Users/xudong/Projects/taichi/python/taichi/lang/common_ops.py", line 172, in assign
    return ti.assign(self, other)
  File "/Users/xudong/Projects/taichi/python/taichi/lang/ops.py", line 79, in wrapped
    return a.element_wise_binary(imp_foo, b)
  File "/Users/xudong/Projects/taichi/python/taichi/lang/matrix.py", line 155, in element_wise_binary
    f'cannot assign scalar expr to '
SyntaxError: cannot assign scalar expr to taichi class {type(a)}, maybe you want to use `a.fill(b)` instead?
```
Aaahhhh, Would you like do give me some guidence about this problem?

Thanks a lot!
Bests,
Xudong

**Comment by yuanming-hu** [comment]
Thanks for the script to reproduce. need to index into `a[i]`:
```python
def test_offset_for_vector():
    offset = 16
    shape = 16
    a = ti.Vector(n=1, dt=ti.i32, shape=shape, offset=offset)


    @ti.kernel
    def test():
        for i in a:
            a[i][0] = 2 * i

    test()
    for i in range(offset, offset+shape, 1):
        assert a[i][0] == 2 * i
    return
```

Let's remove the default value `n=1` in `ti.Vector`. `n` must be specified. Otherwise it is very confusing.

**Comment by Rullec** [comment]
Thanks for your help! I push a new commit to this PR :D. It contains:
1. the offset test for ti.Vector. I tested these functions on my mac and Ubuntu, all passed
2. remove the default value of arg "n=1, n=1, m=1" in both ti.Matrix and ti.Vector
3. replace the ValueError in ti.Matrix and ti.var with AssertionError, and update offset.rst accordly as well.

Would you please check this PR again, Maybe there is still something I can help. :-)

Bests,
Xudong

**Pull request merged**